private/bufpkg/bufcheck/buflint/internal/buflintvalidate/cel.go

```
@@ -0,0 +1,311 @@
+			add(field, celLocation, nil, "cel expression is empty")
+			continue
+		}
+		ast, compileIssues := env.Compile(cel.Expression)
+		switch ast.OutputType() {
+		case types.BoolType, types.StringType, types.ErrorType:
+			// If type is types.ErrorType, compilation has failed and we will
+			// only add the compilation issues.
+		default:
+			add(field, celLocation, nil, "cel expression evaluates to unsupported type: %v", ast.OutputType())
+		}
+		if compileIssues.Err() != nil {
+			for _, parsedIssue := range parseCelIssuesText(compileIssues.Err().Error()) {
+				add(field, celLocation, nil, parsedIssue)
+			}
+		}
+	}
+	return nil
+}
+
+func getReflectMessageDescriptor(resolver protodesc.Resolver, message protosource.Message) (protoreflect.MessageDescriptor, error) {
+	descriptor, err := resolver.FindDescriptorByName(protoreflect.FullName(message.FullName()))
+	if err == protoregistry.NotFound {
+		return nil, fmt.Errorf("unable to resolve MessageDescriptor: %s", message.FullName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	messageDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		// this should not happen
+		return nil, fmt.Errorf("%s is not a message", descriptor.FullName())
+	}
+	return messageDescriptor, nil
+}
+
+func getReflectFieldDescriptor(resolver protodesc.Resolver, field protosource.Field) (protoreflect.FieldDescriptor, error) {
+	descriptor, err := resolver.FindDescriptorByName(protoreflect.FullName(field.FullName()))
+	if err == protoregistry.NotFound {
+		return nil, fmt.Errorf("unable to resolve FieldDescriptor: %s", field.FullName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	fieldDescriptor, ok := descriptor.(protoreflect.FieldDescriptor)
+	if !ok {
+		// this should never happen
+		return nil, fmt.Errorf("%s is not a field", descriptor.FullName())
+	}
+	return fieldDescriptor, nil
+}
+
+func getMessageConstraints(message protosource.Message) (*validate.MessageConstraints, bool, error) {
```

Why not return `*validate.MessageConstraints, error`? It looks like the bool is only true when the message != nil, so it's superfluous, no?

---

There is a case where the message doesn't have this option, but nothing else went wrong, and in this case the helper would return `nil, false, nil`. The caller benefits from knowing whether it should move on (this options isn't present) or it should throw an error (something has gone wrong).

I agree that returning `*validate.MessageConstraints, error` makes its signature cleaner. In that case, it would return some kind of `NotFound` error. I searched for `NotFound` in the codebase and couldn't find a standardized `NotFound` error. Do you have any suggestions?

---

These functions are replaced by direct calls to `DefaultResolver{}.Resolver...`. Resolved in [5e33b02](https://github.com/bufbuild/buf/pull/2473/commits/5e33b02a5044a09ee659acb1d7bde7395490ffe1).